---
title: "Data QC"
---


# Reading

-   [Opportunities in Functional Genomics: A Primer on Lab and Computational Aspects](https://d.pr/f/ngmWkw)


# Objectives

Learn how to take raw RNA sequence data and perform QC

# Introduction

RNA sequencing (RNA-seq) is a powerful technology used to analyze gene expression levels across different conditions or tissues. The raw data generated from RNA-seq experiments is typically stored in fastq files, which contain millions of short sequencing reads that need to be preprocessed before performing differential gene expression analysis. In this tutorial, we will use Kallisto, a fast and accurate transcript quantification tool, to analyze RNA-seq data and identify differentially expressed genes.


# Sequence file format

One common sequence product is a large amount of "short-reads". This often refers to single-end or paired-end reads with a link of \~ 150bp. When data comes off of a sequencer it is often in FASTQ format. This file format contains both sequence and quality informaiton.

A FASTQ file has four line-separated fields per sequence:

-   Field 1 begins with a '\@' character and is followed by a sequence identifier and an optional description (like a FASTA title line).
-   Field 2 is the raw sequence letters.
-   Field 3 begins with a '+' character and is optionally followed by the same sequence identifier (and any description) again.
-   Field 4 encodes the quality values for the sequence in Field 2, and must contain the same number of symbols as letters in the sequence.

https://en.wikipedia.org/wiki/FASTQ_format



## Quality Control

The first step in analyzing RNA-seq data is to perform quality control checks on the raw fastq files. This step is crucial to ensure that the data is of high quality and can be accurately quantified. One popular tool for quality control is FastQC, which generates various quality metrics such as per-base sequence quality, adapter contamination, and GC content.

To perform quality control using FastQC, run the following command:

``` bash
fastqc input.fastq
```

This will generate a HTML report that can be viewed in a web browser.

{{< video https://youtu.be/5LYnq84Pjzk width="400" height="300">}}

Another popular quality control program is `fastp`. [Here](https://sandbox.bio/tutorials?id=fastp-intro) is a very nice tutorial on using `fastp`

