---
title: "hisat"
---

TLDR

# Genome Prep


```{{bash}}
/home/shared/hisat2-2.2.1/hisat2_extract_exons.py \
../data/Amil/ncbi_dataset/data/GCF_013753865.1/genomic.gtf \
> ../output/04-Apulcra-hisat/m_exon.tab
```


```{{bash}}
/home/shared/hisat2-2.2.1/hisat2_extract_splice_sites.py \
../data/Amil/ncbi_dataset/data/GCF_013753865.1/genomic.gtf \
> ../output/04-Apulcra-hisat/m_splice_sites.tab
```


# Build Index

```{{bash}}
/home/shared/hisat2-2.2.1/hisat2-build \
../data/Amil/ncbi_dataset/data/GCF_013753865.1/GCF_013753865.1_Amil_v2.1_genomic.fna \
../output/GCF_013753865.1_Amil_v2.1 \
--exon ../output/04-Apulcra-hisat/m_exon.tab \
--ss ../output/04-Apulcra-hisat/m_splice_sites.tab \
-p 40 \
../data/Amil/ncbi_dataset/data/GCF_013753865.1/genomic.gtf \
2> ../output/04-Apulcra-hisat/hisat2-build_stats.txt
```


# Quantification

```{{bash}}
find /home/shared/8TB_HDD_02/mewing0/clamgonads-macsamples/data/raw/*gz \
| xargs basename -s _R1_001.fastq.gz | xargs -I{} \
/home/shared/hisat2-2.2.1/hisat2 \
-x /home/shared/8TB_HDD_02/mewing0/clamgonads-macsamples/output/hisat/GCF_026571515.1_index \
-p 20 \
-1 /home/shared/8TB_HDD_02/mewing0/clamgonads-macsamples/data/raw/{}_R1_001.fastq.gz \
-2 /home/shared/8TB_HDD_02/mewing0/clamgonads-macsamples/data/raw/{}_R2_001.fastq.gz \
-S ../output/{}.sam
```

# convert SAM to BAM

```{{bash}}
for file in ../output/*sam; do
    base=$(basename "$file" .sam)
    /home/shared/samtools-1.12/samtools view -@ 40 -bS "$file" | \
    /home/shared/samtools-1.12/samtools sort -@ 40 \
    -o ../output/"$base".bam
done
```



# featurecounts
